### ISSUE TITLE
ApolloClientProtocol async/await

### ISSUE THREAD

**Issue by clayellis** [apollographql/apollo-ios#2717]
## Feature request

It would be nice to have first-party support for modern concurrency via Swift `async`/`await` in `ApolloClientProtocol`.

## Motivation

Modern Swift clients have begun to transition away from closure-based APIs to the new Swift Concurrency model. Clients can write their own "migration" layer, but this is prone to error. 

## Proposed solution

Apollo should provide an async variant of each of the methods on `ApolloClientProtocol` (`clearCache`, `fetch`, `watch`, `perform`, `upload`, `subscribe`). A good "first pass" would just include an async variant of `fetch`.

**Comment by calvincestari** [comment]
Thanks @clayellis, you'll be pleased to know that [v2.0 on our roadmap](https://github.com/apollographql/apollo-ios/blob/main/ROADMAP.md#20) is to modernize the whole networking stack in Apollo iOS. We've got a few things still related to 1.0 and code generation to take care of before that but bringing the networking stack up to modern expectations is what we work on next.

I'm going to close this issue as this is already on our roadmap and this alone as "first-party support for modern concurrency" is an overly broad description.

**Comment by clayellis** [comment]
Good to know!

**Comment by clayellis** [comment]
@calvincestari in the meantime — does this appear to be safe?

```swift
extension ApolloClientProtocol {
    func fetch<Query: GraphQLQuery>(
        query: Query,
        cachePolicy: CachePolicy = .default,
        contextIdentifier: UUID? = nil
    ) async throws -> GraphQLResult<Query.Data> {
        return try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<GraphQLResult<Query.Data>, Error>) in
            var cancellable: Cancellable?
            cancellable = self.fetch(query: query, cachePolicy: cachePolicy, contextIdentifier: contextIdentifier, queue: .main) { result in
                do {
                    try Task.checkCancellation()
                } catch {
                    cancellable?.cancel()
                    continuation.resume(throwing: error)
                }

                continuation.resume(with: result)
            }
        }
    }
}
```

**Comment by calvincestari** [comment]
If you haven't already, take a look at https://github.com/apollographql/apollo-ios/issues/2216 which covers this async/await and some of the issues others have experienced with it regarding cache policies.

**Comment by clayellis** [comment]
Oh cool, I wasn't aware of that. Thanks!

**Comment by calvincestari** [comment]
> Sorry to resurrect a closed thread - but what's the path forward for making improvements to the network layer perhaps pre-2.0?
> 
> If changes are possible to introduce additively to the API, will they be accepted into 1.x releases?

Yes I think that is possible but it won't be the focus of our team here at Apollo. If there are community contributions that are robustly tested and introduce some of the async/await features that the community wants then we can help get them reviewed and merged.

> Will 2.0 be in development as perhaps an epic branch in parallel to 1.0, or does everything have to happen end of trunk?

2.0 will probably go the same route as 1.0 did with it's own branch, kept up to date with `main`, so that when it's ready it can be merged back into `main` and everyone gets it as default.
